### ISSUE TITLE
'Cannot assign requested address' error

### ISSUE THREAD

**Issue by lowercasename** [mikebrady/shairport-sync#276]
I'm having trouble getting shairport-sync running on my Raspberry Pi 3 running Jessie. I'm connecting the Raspberry Pi via wifi and have tried using both of our wifi networks. Here's the output of `shairport-sync -v`:

```
Looking for the configuration file "/etc/shairport-sync.conf".
Output device name is "default".
Version: "2.8.2-openssl-Avahi-ALSA-soxr-metadata"
statistics_requester status is 0.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Pan".
Audio Output name is "alsa".
on-start action is "(null)".
on-stop action is "(null)".
wait-cmd status is 0.
mdns backend "avahi".
stuffing option is "0".
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 88 frames.
password is "(null)".
ignore_volume_control is 0.
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range specified by the mixer): 0.
zeroconf regtype is "_raop._tcp".
configuration file name "/etc/shairport-sync.conf" resolves to "/etc/shairport-sync.conf".
metdata enabled is 1.
metadata pipename is "/tmp/shairport-sync-metadata".
get-coverart is 1.
avahi: avahi_register.
avahi: register_service.
avahi: request to add "_raop._tcp" service with metadata
avahi: service 'F6D4A17B27DA@Pan' successfully added.
avahi: address advertised is: "fe80::51e1:6e7e:3fc6:9ca".
avahi: address advertised is: "192.168.15.231".
New RTSP connection on port 5000
Play connection from user agent "AirPlay/280.32.2".
Active-Remote string seen: "2605360973".
DACP-ID string seen: "A008453D17B2EC3F".
Connection from IPv6: fe80::872:2158:a73d:f4c2:59555
Error sendto-ing to timing socket: Cannot assign requested address
Sync packet received before we got a timing packet back.
Error sendto-ing to timing socket: Cannot assign requested address
Error sendto-ing to timing socket: Cannot assign requested address
Sync packet received before we got a timing packet back.
Error sendto-ing to timing socket: Cannot assign requested address
Error sendto-ing to timing socket: Cannot assign requested address
Sync packet received before we got a timing packet back.
Sync packet received before we got a timing packet back.
Sync packet received before we got a timing packet back.
Aliasing of buffer index -- reset.
```
And so forth. Any help would be greatly appreciated!

**Comment by mikebrady** [comment]
Thanks for the report. When you say "both of our networks" could you explain what you mean? Alternatively is there away you could turn IPv6 off temporarily to see if it makes a difference?

**Comment by lowercasename** [comment]
We have two separate internet connections - one is a 4G modem router, one
is (I presume - the setup was not mine) ADSL with a wifi router. I thought
that it might be a router issue, though testing both doesn't necessarily
prove it isn't...

Shall try turning off IPv6 tonight and report back.

>

**Comment by mikebrady** [comment]
Thanks, and sorry for being a bit dense, but is it possible that the 4G router and the ADSL router are both active simultaneously on the same network?

**Comment by mikebrady** [comment]
Hi there. I wonder if I should keep this open, or did you resolve the problem?

**Comment by stevensflight** [comment]
Could you keep it open? I am having the same problem. I tried disabling IPv6 and now I get the following: 

pi@raspberrypi:~ $ shairport-sync -v
Looking for the configuration file "/etc/sha
irport-sync.conf".
Output device name is "default".
Version: "2.8.3-openssl-Avahi-ALSA-metadata"
statistics_requester status is 0.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Shairport Syn
c on raspberrypi".
Audio Output name is "(null)".
on-start action is "(null)".
on-stop action is "(null)".
wait-cmd status is 0.
mdns backend "(null)".
stuffing option is "0".
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 88 frames.
password is "(null)".
ignore_volume_control is 0.
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range
 specified by the mixer): 0.
zeroconf regtype is "_raop._tcp".
configuration file name "/etc/shairport-sync
.conf" resolves to "/etc/shairport-sync.conf
".
metdata enabled is 0.
metadata pipename is "(null)".
get-coverart is 0.
Failed to bind to address 0.0.0.0.
Failed to bind to address ::.
could not bind any listen sockets!
Request to shut down all rtsp conversation t
hreads

**Comment by marcushartwig** [comment]
I don't know if this helps, but I had this issue when my Pi was using only the wifi0 connection. It seemed like shairport-sync was trying to bind to eth0 by default. The solution was to bring eth0 down. After that shairport started normally.

**Comment by mikebrady** [comment]
Thanks for the reports. In my experience, that message – "Failed to bind..." invariably happens when another copy of Shairport Sync is already running, such as happens, for example, if it was started as a system service and you then start it from the command line. I never experienced it happening for any other reason, TBH.

**Comment by mikebrady** [comment]
Actually, I took my own advice and pushed a small update to the `development` branch. Now those error messages read:
```
Unable to listen on IPv4 port 5000. The error is: "Address already in use".
Unable to listen on IPv6 port 5000. The error is: "Address already in use".
Could not listen on port 5000 -- program terminating. Is another Shairport Sync running?
Request to shut down all rtsp conversation threads
asking playing threads to stop
```

**Comment by stevensflight** [comment]
Thanks for the help. I still had no audio with IPv6 disabled. Here is a look at /var/log/syslog from this morning (IPv6 enabled and eth0 at default settings) after a fresh reboot:

May 11 05:58:14 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:17 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:17 raspberrypi rsyslogd-2007: action 'action 1
7' suspended, next retry is Wed May 11 05:58:47 2016 [try h
ttp://www.rsyslog.com/e/2007 ]
May 11 05:58:20 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:23 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:26 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:29 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:32 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:35 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:38 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:41 raspberrypi shairport-sync[546]: Error send
to-ing to timing socket: Cannot assign requested address
May 11 05:58:44 raspberrypi shairport-sync[546]: Error send
:

**Comment by mikebrady** [comment]
Thanks for the report. One bug that's been fixed in the `development` version is this: when a client such as iTunes makes a connection to Shairport Sync, and then when Shairport Sync makes a connection back to the client, say for getting timing information, then, if the Shairport Sync device had multiple IP addresses, it does not necessarily use the same address as the source of the return link as had been used to reach it in the first place.

If the client made a call over IPv6, the connection back would also be over IPv6, but it would not necessarily be in the same scope. I wonder if that isn't causing your problem. It would be easy to check – simply switch over to the development version, which is actually pretty stable.

BTW, I notice that there are `rsyslogd` issues in your log. Do they signify a problem with the Pi's configuration?

**Comment by lowercasename** [comment]
@marcushartwig, bringing eth0 down on the Raspberry Pi worked perfectly!

**Comment by marcushartwig** [comment]
@lowercasename Glad to hear!

My permanent fix was to modify my `/etc/network/interfaces` file to the following. That way I can reboot the Pi as needed and have shairport-sync start automatically.

    auto lo
    iface lo inet loopback

    iface eth0 inet manual
        pre-up ifdown wlan0

    allow-hotplug wlan0
    iface wlan0 inet manual
        wpa-conf /etc/wpa_supplicant/wpa_supplicant.conf
        pre-up ifdown eth0; echo -n '1-1.1:1.0' > /sys/bus/usb/drivers/smsc95xx/unbind
        post-down echo -n '1-1.1:1.0' > /sys/bus/usb/drivers/smsc95xx/bind

**Comment by mikebrady** [comment]
Thanks for the reports. There is definitely a bug in Shairport Sync here when the addresses are link-local IPv6 addresses. I think I have a fix, but it will be another 12 to 24 hours before I push it out.

**Comment by mikebrady** [comment]
I've just pushed a modification that is untested.

**Comment by mikebrady** [comment]
[Update] I think the modification fixes the problem alright. I'd be grateful if you'd try it out. It's in the HEAD of the `development` branch.

**Comment by stevensflight** [comment]
I just installed the development version and it is working.

I am able to stream from my iPhone. But if I turn off AirPlay on the phone and try to stream from my iPad it doesn't work. I can then go back to my phone and have it work. 

Is there a step I am missing to be able to use a different device? 

Syslog shows the following error when trying to stream from my iPad: 

May 14 08:08:29 raspberrypi shairport-sync[539]: Error send
to-ing to timing socket: Cannot assign requested address

**Comment by mikebrady** [comment]
Thanks. Could you let user the full log? It would be good to see the ip numbers used both by the iPhone and the iPad.

**Comment by mikebrady** [comment]
I hope you don't minding me asking an impertinent question, but: Is is possible that you might have compiled the development version but inadvertently used the regular version? It could happen by executing `shairport-sync` rather than `./shairport-sync` in the build directory; I know because I've done it...

You can tell very quickly by the version number information in the log if log verbosity is 1 or more.

**Comment by stevensflight** [comment]
Sorry I haven't replied sooner. This is a low priority project for me. I will see if I can play with it some in the next couple of days.

**Comment by retrostrobe** [comment]
@marcushartwig's comment about bringing eth0 down worked for me. Editing the interfaces file didn't do the trick, so I edited `/etc/rc.local` and added: `(sleep 30;sudo ifconfig eth0 down)&` just before `exit 0`.

**Comment by stevensflight** [comment]
@retrostrobe - that worked for me too. I can stream from my phone, switch to my iPad, and back and it is still working!  Thanks!

**Comment by mikebrady** [comment]
Thanks for this. I've been trying to fix it so that this workaround isn't necessary, but it's good that it works!

**Comment by mikebrady** [comment]
I believe this is fixed by 2.8.4. Feel free to reopen if not. Thanks.
